### ISSUE TITLE
Fixed: Fatal error when submit an empty required date field with popup widget.

### ISSUE THREAD

**Issue by argiepiano** [backdrop/backdrop-issues#6275]
## Description of the bug

Trying to create or edit a node that contains a required date field (popup widget, date and time granularity) results in a fatal error:

`TypeError: date_offset_get(): Argument #1 ($object) must be of type DateTimeInterface, null given in date_offset_get() line 1703 of /Users/XXX/Sites/localhost/XXX/core/modules/date/date.elements.inc` 

## Steps To Reproduce

To reproduce the behavior:
1. Attach a date field to a content type. Granularity up to minutes. Widget: popup
2. Make it required
3. Visit the content creation form
4. Leave the date field blank and submit the form

## Actual behavior

Fatal error occurs.

## Expected behavior

Validation should prevent the form submission - without a fatal error.

## Additional information

- Backdrop CMS version: 1.26.1
- Web server and its version:  Apache/2.4.57 (Debian)
- PHP version:  8.1.24

(Reproduced both locally and in a sandbox)

**Comment by argiepiano** [comment]
This error is fixed in D7 7 years ago, by [this commit](https://git.drupalcode.org/project/date/-/commit/c8771964ab47385490996ed4a86ae6adc2979ef6), which was created after a (somewhat related) error reported in issue [1889744](https://www.drupal.org/project/date/issues/1889744).

This issue was included in a "todo" list in Backdrop back in 2019 in #4056, and also in #4057. None of those tasks have been completed - so we still have a pretty buggy date field! I'll try to submit a PR for this one...

**Comment by argiepiano** [comment]
Interesting problem: the patch I linked above solves this issue, BUT not all possible related issues. The current version of D7's date still has a fatal error if you have required "From" and "To" dates, and you leave the "From" date empty. 

So, I could submit a literal port of the D7 fix, but it won't solve that related issue. Question: should I try to solve both this, and related issues here? Open a new bug report once this one is fixed? 

There is a potential fix for both issues, but it hasn't been merged in D7: [2933157](https://www.drupal.org/project/date/issues/2933157). I could try to combine both the merged one and this one.

**Comment by bugfolder** [comment]
Seems to me like fixing all the ways of failure with empty date fields at once would be appropriate.

**Comment by argiepiano** [comment]
PR https://github.com/backdrop/backdrop/pull/4552 provided.

The PR borrows heavily from the two D7 fixes posted above. While they look complicated, they are fairly straightforward: the rearrange some `if` blocks.

The PR also adds some automated tests to check some of the situations that generated fatal errors.

To test
1. Create a required date field (popup) and save node without entering a value (fatal before fixing on PHP 8.1)
2. Create a required date field with an end date, and save the node by providing an end date but not a start date (fatal too)

There are a couple of coding standard failures - but those are doc blocks for code this PR hasn't touched. All test classes in `date` have the same issue: no doc blocks. So rather than adding one to that class, I decided to ignore it.

**Comment by klonos** [comment]
Thanks for working on this @argiepiano 🙏🏼 

I haven't tested this yet, but did a quick code review. Only some nits were picked.

**Comment by argiepiano** [comment]
Thanks, @klonos. I've adopted your suggestions. Open to suggestions regarding one of your comments, though.

**Comment by bugfolder** [comment]
Did some testing, found another crashing situation.

* Create a date field on your node (Field type = "Date (Unix timestamp)", Widget = "Popup calendar")
* Make it required
* In Field Settings, check "Collect an end date", but don't make it required.

Now, create your test node. Check the "Show end date" checkbox, but leave it empty. Save. You'll see something like this:

![image](https://github.com/backdrop/backdrop-issues/assets/29214084/a2fb35d4-2b58-416b-bce6-dd950917c8a7)

Interestingly, if you don't check the "Show end date" checkbox, it saves with no problems and the end date is blank.

**Comment by argiepiano** [comment]
> Now, create your test node. Check the "Show end date" checkbox, but leave it empty. Save. You'll see something like this:

Revisiting this PR. Just FTR, before this PR, when clicking "Show end date" for a popup date field that does not require the end date, but does require  starting date, you get `The end date must be greater than the start date.`.

**Comment by argiepiano** [comment]
@bugfolder, I wanted to post more details about the current behavior (without the PR) for dates with **required** start dates, and **non-required** end dates.  

- If you click "Show end date", Backdrop will not allow you to submit a blank end date. You get the validation error about the end date needing to be greater than the start date. In fact, if your field has the "Collect an end date" setting on, you cannot submit the form when the date end field is shown, unless you enter a value there. 
- When you don't click "Show end date", Backdrop will actually enter an end date for you - the same date as the start date. This is true even if the end date is not required.

So, in a nutshell, if you chose "Collect an end date" and you click "Show end date" in the form, the end date becomes required. Whether this is a good idea or not, is really beyond the scope of this issue, which focused on an error produced by submitting a form with an empty required date with popup widget.

So, I will modify the current PR to behave the same way it currently does re: empty end date. We can open another issue to address what I described above.

**Comment by argiepiano** [comment]
Modified PR ready for testing. Testing directions in the OP.

**Comment by argiepiano** [comment]
Updated PR is now ready for testing and code review. Tests are all passing. PHPCS indicates a problem with code this PR hasn't touched. 

This PR fixes the OP (fatal error when submitting a required popup date field), and also fixes the problem pointed out [above](https://github.com/backdrop/backdrop-issues/issues/6275#issuecomment-1821145779) with fatal errors when the start date was left empty for a required date combo field.

As I mentioned above, this PR preserves some potentially problematic behavior when you use a date field with a "non-required" end date (i.e. having set "Collect an end date" to TRUE for the field).  Currently, and also before this PR, the form will not allow you to submit the date if you click "Show end date" and leave the end date blank (even though the end date is apparently "not required" in the field definition). I feel this should be addressed in a separate issue.
 
@bugfolder, @klonos I'll be thankful if you could test the new PR.

**Comment by argiepiano** [comment]
@bugfolder, hi! Wondering if you might have the time to test the latest PR for this nasty Date bug?

**Comment by bugfolder** [comment]
Tested.

* Create a date field on your node (Field type = "Date (Unix timestamp)", Widget = "Popup calendar")
* Make it required
* In Field Settings, check "Collect an end date", but don't make it required.
* Leave fields blank.
* Submit.

Result is correct: submission error with "A valid date is required for Date popup with end start date." (This formerly crashed.)

Leave date field blank (the OP scenario):

Result is correct, submission error as expected.

So the scenarios described above that result in crashes are fixed by the PR.

I do have one question, but it may be out-of-scope for this issue. If I specify a Date granularity of Hour and minute and make the field required, both the date and time fields are shown on the form, but I can successfully submit the form by leaving the time field blank. It seems like if the field is required, both sub-fields (date and time) should be required. Indeed, if I leave the date field blank, both subfields get the red highlight:

![image](https://github.com/backdrop/backdrop-issues/assets/29214084/963c9f0e-d1f5-4830-9ebb-b58a86536b84)

That led me to try leaving the date field blank but filling in the time. Lo and behold, another crash:

![image](https://github.com/backdrop/backdrop-issues/assets/29214084/29e2ab40-e561-4c77-aeca-f03c642723d3)

So still needs work, I guess.

**Comment by alanmels** [comment]
The fix in PR worked well for me too.

**Comment by argiepiano** [comment]
Thanks @alanmels. I intend to come back to this issue to fix the additional problem @bugfolder found. Soon. Sorry it's taking so long.

**Comment by argiepiano** [comment]
@herbdool I noticed you removed the milestone. I don't understand the reason for this. I've actually made the fixes asked by @bugfolder. Would you mind adding the milestone back?

@bugfolder after a long while I came back to this.
1. I solved the fatal error produced by submitting a time but not a date.
2. As to your comment "but I can successfully submit the form by leaving the time field blank" - can we open a separate issue for that? The goal of this was to prevent a fatal error when submitting a required date field.

Would you be able to test again? Let's fix this nasty bug.

The PR involves quite a bit of rewriting. Some of that has been inspired by changes made in D7 some time ago that never made it into Backdrop.

About the automated tests:
- The spelling failures are meaningless to me. Not sure if I can to anything about this
- The coding standard failure seems to refer to code this PR did not touch. Let me know otherwise

Thanks!

**Comment by herbdool** [comment]
Anything which looked like it was stagnating and wouldn't make it into a bug release soon, I set back to milestone candidate. Otherwise the milestone tends to just get larger and larger and it's hard to find what is worth wrapping up. 

But now that you've put it back to needs testing and review, I'll put it back.

**Comment by stpaultim** [comment]
I did some testing and it works for me. I tested the original issue and the issue described by Bugfolder a few comments back.

**Comment by laryn** [comment]
Thank you @argiepiano for filing and seeing this one through. Thanks for your thorough work reviewing and testing @bugfolder!
And thanks for testing and reviews by @stpaultim, @klonos, @herbdool and @alanmels.
Also with thanks to tatyana, mstrelan, mariagwyn for their work in Drupal 7.
Merged into `1.x` and `1.27.x`.
